Sources/NIOHTTPResponsiveness/HTTPDrippingDownloadHandler.swift

```diff
@@ -0,0 +1,228 @@
+            DrippingState(
+                chunksLeft: self.count,
+                currentChunkBytesLeft: self.size
+            )
+        )
+
+        self.writeChunk(context: context)
+    }
+
+    public func channelInactive(context: ChannelHandlerContext) {
+        self.phase = .done
+        self.scheduled?.cancel()
+    }
+
+    public func channelWritabilityChanged(context: ChannelHandlerContext) {
+        if case .dripping = self.phase, self.pendingWrite && context.channel.isWritable {
+            self.writeChunk(context: context)
+        }
+    }
+
+    private func writeChunk(context: ChannelHandlerContext) {
+        // Make sure we don't accidentally reenter
+        if self.activelyWritingChunk {
+            return
+        }
+        self.activelyWritingChunk = true
+        defer {
+            self.activelyWritingChunk = false
+        }
+
+        // If we're not dripping the response body, there's nothing to do
+        guard case .dripping(var drippingState) = self.phase else {
+            return
+        }
+
+        // If we've sent all chunks, send end and be done
+        if drippingState.chunksLeft < 1 {
+            self.phase = .done
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+            return
+        }
+
+        var dataWritten = false
+        while drippingState.currentChunkBytesLeft > 0, context.channel.isWritable {
+            let toSend = min(drippingState.currentChunkBytesLeft, HTTPDrippingDownloadHandler.downloadBodyChunk.readableBytes)
+            let buffer = HTTPDrippingDownloadHandler.downloadBodyChunk.getSlice(
+                at: HTTPDrippingDownloadHandler.downloadBodyChunk.readerIndex,
+                length: toSend
+            )!
+            context.write(self.wrapOutboundOut(.body(buffer)), promise: nil)
+            drippingState.currentChunkBytesLeft -= toSend
+            dataWritten = true
+        }
+        if dataWritten {
+            context.flush()
+        }
+
+        // If we weren't able to send the full chunk, it's because the channel isn't writable. We yield until it is
+        if drippingState.currentChunkBytesLeft > 0 {
+            self.pendingWrite = true
+            self.phase = .dripping(drippingState)
+            return
+        }
+
+        // We sent the full chunk. If we have no more chunks to write, we're done!
+        drippingState.chunksLeft -= 1
+        if drippingState.chunksLeft == 0 {
+            self.phase = .done
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+            return
+        }
+
+        // More chunks to write.. Kick off timer
+        drippingState.currentChunkBytesLeft = self.size
+        self.phase = .dripping(drippingState)
+        let this = NIOLoopBound(self, eventLoop: context.eventLoop)
+        let loopBoundContext = NIOLoopBound(context, eventLoop: context.eventLoop)
+        self.scheduled = context.eventLoop.scheduleTask(in: self.frequency) {
```

Comment:
Thanks for the suggestion. I _think_ I used it mostly correctly, but I'm not 100% sure. I haven't built much intuition around when swift does or doesn't allocate. I did find the fact that `scheduleCallback` throws a bit surprising and looked for an equivalent syncOperations-like method to avoid the `try`, but I didn't find one. I think `try!` may be fine in practice given current implementations, but not entirely correct. Interested in your feedback here.